### PR TITLE
Load Gfxobjects

### PR DESCRIPTION
**Pull request by Nemrav**
Very very in progress define loading for gfx objects.
- Lots of verifying to do (int64 vs fixed_point_t, are things correctly marked case sensitive vs insensitive, are properties EXACTLY_ONE or ZERO_OR_ONE, etc.)
- Need to reorganize/improve some datatypes (extraction of format_t to a datatype, improve unlabelled vec3 and vec4 types).

### CODE REVIEW

File: src/openvic-simulation/interface/GFXObject.cpp [hunk 1]
```diff
@@ -190,3 +136,178 @@ bool Actor::_fill_key_map(NodeTools::case_insensitive_key_map_t& key_map) {
...
+	no_of_frames { 1 }, font_size { 7 }, font {} {}
+
+//TODO: Verify font_size is int
+bool UnitStatsBillboard::_fill_key_map(NodeTools::case_insensitive_key_map_t& key_map) {
+	bool ret = Object::_fill_key_map(key_map);
+
+	ret &= add_key_map_entries(key_map,
+		"textureFile", ONE_EXACTLY, expect_string(assign_variable_callback_string(texture_file)),
+		"mask", ONE_EXACTLY, expect_string(assign_variable_callback_string(mask_file)),
+		"effectFile", ONE_EXACTLY, expect_string(assign_variable_callback_string(effect_file)),
```
Comment (Hop311) [inline src/openvic-simulation/interface/GFXObject.cpp:261]:
```suggestion
		"effectFile", ONE_EXACTLY, expect_string(assign_variable_callback_string(effect_file)),
```
Two stray tab characters at the end of this line

File: src/openvic-simulation/interface/GFXObject.cpp [hunk 2]
```diff
@@ -190,3 +136,178 @@ bool Actor::_fill_key_map(NodeTools::case_insensitive_key_map_t& key_map) {
...
+	ret &= add_key_map_entries(key_map,
+		"textureFile", ZERO_OR_ONE, expect_string(assign_variable_callback_string(texture_file)),
+		"scale", ZERO_OR_ONE, expect_fixed_point(assign_variable_callback(scale))
+	);
+
+	return ret;
+}
+
+/* MapItems.gfx */
+Projection::Projection() :
```
Comment (Hop311) [inline src/openvic-simulation/interface/GFXObject.cpp:201]:
```suggestion
Projection::Projection() :
```
Trailing space at the end of this line, same for `Billboard`, `UnitStatsBillboard`, `TextBlock` and `AnimatedMapText`'s constructors

File: src/openvic-simulation/interface/GFXObject.hpp [hunk 3]
```diff
@@ -74,4 +77,232 @@ namespace OpenVic::GFX {
...
+	};
+
+	/* Core.gfx */
+	class AnimatedMapText final : public Object {
+		friend std::unique_ptr<AnimatedMapText> std::make_unique<AnimatedMapText>();
+
+		using enum text_format_t;
+
+	public:
+		struct TextBlock {
```
Comment (Hop311) [inline src/openvic-simulation/interface/GFXObject.hpp:271]: Unless there can be multiple text blocks, might as well just have its contents as regular `AnimatedMapText` variables without the intermediate struct

**Review by Hop311 — COMMENTED** [review]
Amazing work, very happy to see you expand the interface loading system so competently 👍

File: src/openvic-simulation/interface/GFXObject.cpp [hunk 4]
```diff
@@ -190,3 +136,178 @@ bool Actor::_fill_key_map(NodeTools::case_insensitive_key_map_t& key_map) {
...
+		"duration", ZERO_OR_ONE, expect_fixed_point(assign_variable_callback(duration)),
+		"fadeout", ZERO_OR_ONE, expect_fixed_point(assign_variable_callback(fadeout))
+	);
+
+	return ret;
+}
+
+
+Billboard::Billboard() : 
+	texture_file {}, scale { 1 }, no_of_frames { 1 }, font_size { 7 },
```
[the remark this answers is not in the document]
Comment (Nemrav) [reply src/openvic-simulation/interface/GFXObject.cpp:232]: - noOfFrames = 0 causes vic2 to crash, the default really does appear to be 1 based on the capital not having the field by default.
- UnitStatsBillboards aren't used in game as of HoD (commenting them out doesn't do anything, textures they reference appear to be from the base game, pre-expansion), so I think they can be commented out.

[the remark this answers is not in the document]
Comment (Nemrav) [reply src/openvic-simulation/interface/GFXObject.cpp:232]: The font_size, offset2, and font properties of billboardType are only present in billboards unused in vic2 (they appear to be hoi3 leftovers). Shall I remove those, replacing them with success_callback?

[the remark this answers is not in the document]
Comment (Nemrav) [reply src/openvic-simulation/interface/GFXObject.cpp]: I replaced those unused properties with success_callback.

**Review by Hop311 — APPROVED** [review]
Great work!

**Pull request merged**